### ISSUE TITLE
Change Request: Provide a better system for "extension" rules

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [eslint/eslint#19173]
### ESLint version

9.15.0

### What problem do you want to solve?

Forking out of #19134 & https://github.com/typescript-eslint/typescript-eslint/issues/10338 -> https://github.com/eslint/eslint/issues/19169#issuecomment-2499150529:

> The only acceptable way to extend a core rule is to copy the file into your own project and then wrap it however you want. That does put more maintenance burden on plugin developers, but that's the point. If you want to do something that isn't officially supported, you need to take full responsibility for that implementation.

This is not an ideal situation for plugin developers. The required extra work in lieu of having a pluggable API or abstractions necessitates that plugin developers:

* Maintain automations to keep up with ESLint's releases - including copying rules and any imported dependencies from ESLint core & its package dependencies
    * Amusingly, this again leads plugins to rely on implementation details. Just, now that would only be during their build steps, not their published packages
* Be deeply familiar with and having a dependency on the potentially any/all implementation details of base rules

Much of that work is already in play today, including taking the dependency on base rules. Both the current unsupported approach _and_ the recommended _"copy and wrap"_ approach cause higher-than-expected levels of implementation detail reliance in extensions.

### What do you think is the correct solution?

I honestly have no good idea. If the goal of ESLint is to not allow plugins to rely on _any_ behavior from rules -which is my interpretation of https://github.com/eslint/rfcs/pull/80#discussion_r629623105- then generalizing any of the following strategies we typically do in typescript-eslint seem moot.

For example, [`@typescript-eslint/no-unused-expressions`](https://github.com/typescript-eslint/typescript-eslint/blob/f5e23e28653705fc7021d8c780f0d63656dea518/packages/eslint-plugin/src/rules/no-unused-expressions.ts) effectively wraps the base rule's `ExpressionStatement` with filtering logic. But if we can't rely on the base rule having any particular node selectors, then... 🤷

### Participation

- [x] I am willing to submit a pull request for this change.

### Additional comments

Personally, I don't think resolving this issue should block #19169. The _"don't do this"_ recommendation with the current _"copy and paste"_ approach can be added regardless of this issue. I suspect finding a more preferable strategy (this issue) will take much longer. If and when an alternate strategy is resolved, it can always add it to the docs.

I'll also note that plugins aren't "choosing" to write extension rules: they _have_ to to satisfy user need. For example, typescript-eslint's extension rules must act the same as their base rules _except_ for cases where TypeScript's syntax and/or type checking change something. I see extension rules as an expected use case for plugins that just hasn't yet had a formalized first-party ESLint API. If plugins could avoid taking any dependency on core rule behavior in this way -and therefore inconveniencing the ESLint project- that would be even better!

**Comment by fasttime** [comment]
Thanks for this issue @JoshuaKGoldberg. I'll just drop here some links to two existing discussions/issues that seem to contain interesting ideas for the current discussion, albeit originating from a different motivation.

* #18830 contains a suggestion to update core rules so that they become TypeScript-aware. If implemented, I think that this suggestion could benefit `typescript-eslint` in a number of ways. For example, currently most built-in rules are only tested against JavaScript syntax, meaning that they could provide incorrect results/autofix/suggestions for TypeScript code.
If TypeScript support becomes official, this incorrect behavior will have to be fixed in the base rule rather than in the extension. Also, rules that still need `typescript-eslint` extensions for type awareness could delegate linting to their underlying rules in cases that don't need special handling, regardless of TS syntax (As noted in the issue, this is currently not safe because implementation details of a rule could change anytime).

* There are plans to move built-in rules to the `@eslint/js` plugin in a future major release (tracked by #19013). If that happens, plugins that extend core rules could pin the version of `@eslint/js` in their dependencies, and that will keep them safe from implementation changes in the underlying rules while still maintaining compatibility with new versions of `eslint`.
Note that #17656 was not considered a breaking change because it was implemented entirely inside ESLint and `@eslint/eslintrc`. If another similar change happens, it will require changes to the rules in language plugins like `@eslint/js` as well, and those changes to plugin rules will be considered breaking because they will make the `@eslint/js` plugin incompatible with previous (minor) releases of ESLint.

**Comment by nzakas** [comment]
> I honestly have no good idea. If the goal of ESLint is to not allow plugins to rely on _any_ behavior from rules -which is my interpretation of [eslint/rfcs#80 (comment)](https://github.com/eslint/rfcs/pull/80#discussion_r629623105) - then generalizing any of the following strategies we typically do in typescript-eslint seem moot.

You got it. Just to make sure I'm 100% clear:

- Core rules are *not* part of the public API
- Rules were not designed to allow inheritance
- Extending core rules by loading them at runtime from `eslint` is *not* a supported use case
- Extending any rules at runtime from any plugin is incredibly fragile
- We do *not* guarantee that the rules will continue to work in such a way that anyone extending them won't break
- There are no plans to encourage or support this behavior in any way
- Core rules *will* be removed from the `eslint` package at some future point (moved to `@eslint/js`)

Now, once rules are moved to `@eslint/js`, they can be reliably pulled out of the plugin object via the `rules` key. Does that mean it's safe to extend them from there? **No.** Because we still can't guarantee that rules won't change in such a way that it won't break rules that are built on top of them. 

And, once again, for the record: The *only* safe way to extend rules you don't own is to copy the source file over. That's the only way you can guarantee that your rule won't break when the rule-owner package is updated. Otherwise you are rolling the dice that a user is using only a compatible version of the rule-owner package.

(Note: This is actually a fairly common computer science problem. Extending a class that you have no control over, i.e., a class you've imported from another package, invites breakages when the package is updated *unless* the class is specifically designed to be subclassable in this way.)
---
With all of that said, if you still choose to pursue this way of developing rules, you're voiding your warranty and taking all responsibility for possible breakages onto yourself.

**Comment by JoshuaKGoldberg** [comment]
> With all of that said, if you still choose to pursue this way of developing rules, you're voiding your warranty and taking all responsibility for possible breakages onto yourself.

Confirmed, we don't want to pursue this way of developing rules. I think we're all in alignment that the "extension" rules system is problematic. Even if core rules were part of the public API (ACK, they're very much not!), we would be asking for something that wouldn't require either copy & pasting or relying on core rule internals. But, confirmed, _core rules are not part of the API_ and we're not asking to make them part of it.

💡 I just renamed the issue to indicate it's asking for a better system _than_ "extension" rules, rather than _for_.

All that being said, many ESLint core rules are broken in other languages, i.e. TypeScript. The root reason why "extension" rules came to be is still present. Can we discuss finding a system that improves how plugins -i.e. typescript-eslint- fixes for that core need?

Personally, I see two levels to solve:

1. Syntax-level: for pairs like [`default-param-last`](https://eslint.org/docs/latest/rules/default-param-last) -> [`@typescript-eslint/default-param-last`](https://typescript-eslint.io/rules/default-param-last) where the core ESLint rule is broken on TypeScript _syntax_
2. Type-level: for pairs like [`prefer-promise-reject-errors`]() -> [`@typescript-eslint/prefer-promise-reject-errors`](https://typescript-eslint.io/rules/prefer-promise-reject-errors) where type information is used to make the rule more capable

+1 to https://github.com/eslint/eslint/issues/19173#issuecomment-2503570112 that TypeScript support becoming official would solve level 1 for TypeScript, and adding in a built-in pluggable type system would solve level 2 for TypeScript. The specific use case of typescript-eslint would be satisfied. But, I don't know if there are other "extension" rules out there that would want to be replaced with another system.

**Comment by absidue** [comment]
I think the rules in the first category could potentially be solved in the JavaScript rules by adding support for type-stripping i.e. not fully type aware but type tollerant, so that if you pass a TypeScript file it, it will be able to ignore the TypeScript syntax and treat it like a JavaScript file. But as that will likely increase the maintenance burden for the ESLint team a lot, having to keep up with new TypeScript versions and syntax, I think that type-stripping support in the JavaScript rules should be considered community maintained.
What do I mean by that:
1. The ESLint team themselves does not have to take any responsibility for breakages caused by TypeScript syntax as those rules are first and foremost for JavaScript, that also means that breakages to the type-stripping would be okay even in minor releases. This should probably be explicity documented so that users can align the expectations appropriately
2. The ESLint team could allow community members to submit pull requests that add type-stripping support, maybe the typescript-eslint team could take on that responsiblity
3. If there are any discrepances in behaviour between JavaScript and TypeScript language features e.g. decorators, the JavaScript version should be the supported version

The second category that are rules that are written specifically for TypeScript, so just like the JSON and CSS rules the correct approach would still be to copy them into the typescript-eslint project and maintain the TypeScript versions of them there.

**Comment by nzakas** [comment]
@absidue type-stripping isn't really an option because ESLint itself isn't parsing the file, typescript-eslint is, and it's producing an AST with the type information in it. If we cared only about getting core rules to work, yes, we could create a parser that just stripped the types out of TypeScript files and we would (mostly) have things working. But given that the typescript-eslint rules want that extra type information, we'd actually be moving further away from the desired end state.

> All that being said, many ESLint core rules are broken in other languages, i.e. TypeScript. The root reason why "extension" rules came to be is still present. Can we discuss finding a system that improves how plugins -i.e. typescript-eslint- fixes for that core need?

Right, and I would argue it's expected that core rules would not work in other languages. After all, how is ESLint supposed to properly anticipate unknown syntax to make the rule work? I would expect that anything that isn't pure JavaScript would cause some or all rules not to work.

I agree that the only real solution to category 1 is to have the core rules understand TypeScript syntax. As [I previously stated](https://github.com/eslint/eslint/discussions/18830), I'm amenable to that. I *think* that's your real concern, getting core rules to work correctly with TypeScript (rather than any "other language"). The question here would be how much work that is? Are we just talking checking for TypeScript nodes? Or would we need eslint-scope changes too?

For category 2, I'm not sure how solvable that is. Because ESLint can't operate cross-files, any type information we could get in the core would be limited to what's in the file. So could we figure out if something was a promise? In most cases yes. Could we figure it out if it were imported from another file? No.

I'm inclined to argue that category 2 is a good example of when it makes sense to just create an entirely new rule in typescript-eslint.

**Comment by fasttime** [comment]
Based on the discussion above and in the TSC meeting, we'll keep this issue open for tracking until there is a resolution on #18830.

**Comment by JoshuaKGoldberg** [comment]
> your real concern, getting core rules to work correctly with TypeScript (rather than any "other language")

Well, it's my immediate real concern. But I'm under the impression this will be an issue once a competitor to TypeScript like [Ezno](https://github.com/kaleidawave/ezno) gets its own plugin.

Separately, https://github.com/sindresorhus/eslint-plugin-unicorn/issues/2496 makes me suspect it's not just a languages concern. `unicorn/expiring-todo-comments` is in a sense an "extension" rule around `no-warning-comments`. The same systemic issues with "extension" rules apply: I'd think the proposed solution of copy & pasting `no-warning-comments` unwieldy and causing effort on `eslint-plugin-unicorn`'s part. (cc @sindresorhus as fyi)

**Comment by nzakas** [comment]
> Well, it's my immediate real concern. But I'm under the impression this will be an issue once a competitor to TypeScript like [Ezno](https://github.com/kaleidawave/ezno) gets its own plugin.

Yeah, and as I said, I think assuming that all core JS rules will work with any JS derivative regardless of syntax differences is an unrealistic expectation. I don't see this as a generally-solvable problem.

I feel like we are just going in circles at this point. Extending core (or any other) rules is a bad practice, not one we want to encourage nor outright support. If you want to do it, you live with the consequences and the fragility that comes with it. You might feel like copy-pasting rules is too heavy, but again, when you are doing something that is not officially supported, it's up to you to bear such costs.

**Comment by nzakas** [comment]
The only question left on this issue is if we want to make core rules TypeScript-aware to avoid duplication with typescript-eslint. What do people think of that?

**Comment by JoshuaKGoldberg** [comment]
I'm obviously very biased here, but am in favor. 🙂 

Surfacing an earlier note: this wouldn't completely eliminate the duplication / extension rules, but would clear out a large swathe of them.

**Comment by nzakas** [comment]
> Surfacing an earlier note: this wouldn't completely eliminate the duplication / extension rules, but would clear out a large swathe of them.

Yup, understood. I think any incremental progress we can make is beneficial. No need to let the perfect be the enemy of the good.

**Comment by fasttime** [comment]
> The only question left on this issue is if we want to make core rules TypeScript-aware to avoid duplication with typescript-eslint. What do people think of that?

Personally I'm in favor of adding TypeScript syntax support to built-in rules, and I agree that rules requiring type-awareness to work with TypeScript code are better suited to typescript-eslint. This is being discussed in #18830 but I haven't seen much activity on that discussion lately.

**Comment by nzakas** [comment]
**TSC Summary:** This issue proposes starting to make ESLint core rules TypeScript-aware in order to eliminate duplication with typescript-eslint.

**TSC Question:** Do we want to accept this proposal?

**Comment by nzakas** [comment]
@JoshuaKGoldberg @bradzacher we'd like to move forward with making some core rules TypeScript syntax-aware. Do you have a list of rules that would be appropriate as a "getting started" list?

**Comment by JoshuaKGoldberg** [comment]
Fantastic! From the filters on <https://typescript-eslint.io/rules>:

* [All extension rules](https://typescript-eslint.io/rules/?=extension): 25 rules
* 👉 [Syntax-only extension rules](https://typescript-eslint.io/rules/?=extension-xtypeInformation): 7 rules
* [Type-checked extension rules](https://typescript-eslint.io/rules/?=extension-typeInformation): 18 rules

I think the 👉 syntax-only extension rules list is relevant here. That's:

- [`@typescript-eslint/class-methods-use-this`](https://typescript-eslint.io/rules/class-methods-use-this)
- [`@typescript-eslint/default-param-last`](https://typescript-eslint.io/rules/default-param-last)
- [`@typescript-eslint/init-declarations`](https://typescript-eslint.io/rules/init-declarations)
- [`@typescript-eslint/max-params`](https://typescript-eslint.io/rules/max-params)
- [`@typescript-eslint/no-array-constructor`](https://typescript-eslint.io/rules/no-array-constructor)
- [`@typescript-eslint/no-dupe-class-members`](https://typescript-eslint.io/rules/no-dupe-class-members)
- [`@typescript-eslint/no-empty-function`](https://typescript-eslint.io/rules/no-empty-function)
- [`@typescript-eslint/no-invalid-this`](https://typescript-eslint.io/rules/no-invalid-this)
- [`@typescript-eslint/no-loop-func`](https://typescript-eslint.io/rules/no-loop-func)
- [`@typescript-eslint/no-loss-of-precision`](https://typescript-eslint.io/rules/no-loss-of-precision)
- [`@typescript-eslint/no-magic-numbers`](https://typescript-eslint.io/rules/no-magic-numbers)
- [`@typescript-eslint/no-redeclare`](https://typescript-eslint.io/rules/no-redeclare)
- [`@typescript-eslint/no-restricted-imports`](https://typescript-eslint.io/rules/no-restricted-imports)
- [`@typescript-eslint/no-shadow`](https://typescript-eslint.io/rules/no-shadow)
- [`@typescript-eslint/no-unused-expressions`](https://typescript-eslint.io/rules/no-unused-expressions)
- [`@typescript-eslint/no-unused-vars`](https://typescript-eslint.io/rules/no-unused-vars)
- [`@typescript-eslint/no-use-before-define`](https://typescript-eslint.io/rules/no-use-before-define)
- [`@typescript-eslint/no-useless-constructor`](https://typescript-eslint.io/rules/no-useless-constructor)

**Comment by nzakas** [comment]
> I think the 👉 syntax-only extension rules list is relevant here. That's:

That seems like a good place to start. :+1: I've turned that into a checklist below:

- [ ] [`@typescript-eslint/class-methods-use-this`](https://typescript-eslint.io/rules/class-methods-use-this)
- [ ] [`@typescript-eslint/default-param-last`](https://typescript-eslint.io/rules/default-param-last)
- [ ] [`@typescript-eslint/init-declarations`](https://typescript-eslint.io/rules/init-declarations)
- [ ] [`@typescript-eslint/max-params`](https://typescript-eslint.io/rules/max-params)
- [ ] [`@typescript-eslint/no-array-constructor`](https://typescript-eslint.io/rules/no-array-constructor)
- [ ] [`@typescript-eslint/no-dupe-class-members`](https://typescript-eslint.io/rules/no-dupe-class-members)
- [ ] [`@typescript-eslint/no-empty-function`](https://typescript-eslint.io/rules/no-empty-function)
- [ ] [`@typescript-eslint/no-invalid-this`](https://typescript-eslint.io/rules/no-invalid-this)
- [ ] [`@typescript-eslint/no-loop-func`](https://typescript-eslint.io/rules/no-loop-func)
- [ ] [`@typescript-eslint/no-loss-of-precision`](https://typescript-eslint.io/rules/no-loss-of-precision)
- [ ] [`@typescript-eslint/no-magic-numbers`](https://typescript-eslint.io/rules/no-magic-numbers)
- [ ] [`@typescript-eslint/no-redeclare`](https://typescript-eslint.io/rules/no-redeclare)
- [ ] [`@typescript-eslint/no-restricted-imports`](https://typescript-eslint.io/rules/no-restricted-imports)
- [ ] [`@typescript-eslint/no-shadow`](https://typescript-eslint.io/rules/no-shadow)
- [ ] [`@typescript-eslint/no-unused-expressions`](https://typescript-eslint.io/rules/no-unused-expressions)
- [ ] [`@typescript-eslint/no-unused-vars`](https://typescript-eslint.io/rules/no-unused-vars)
- [ ] [`@typescript-eslint/no-use-before-define`](https://typescript-eslint.io/rules/no-use-before-define)
- [ ] [`@typescript-eslint/no-useless-constructor`](https://typescript-eslint.io/rules/no-useless-constructor)
- [ ] [core `no-var`](https://github.com/typescript-eslint/typescript-eslint/issues/2594)

Anyone want to volunteer to work on one or more of these?

**Comment by JoshuaKGoldberg** [comment]
Switching back from https://github.com/eslint/eslint/issues/19373#issuecomment-2619979489: I'm not sure we're aligned on what _"TypeScript syntax-aware"_ means here. What are the kinds of changes marked as accepted for these rules?

Taking [`@typescript-eslint/no-unused-vars`](https://typescript-eslint.io/rules/no-unused-vars) as an example:

* Are type-only constructs -`interface A { }`, `type A = {}`- to be checked?
* How about TypeScript-only runtime constructs: `enum A {}`, `namespace A {}`?
* Will ESLint recognize the `declare` in `declare const a;`?

I'd assumed _"TypeScript syntax-aware"_ meant adding TypeScript-specific AST concepts such as `enum` and `interface` nodes to the AST. That would mean all three of those example cases are supported by the core rules, and the extension rules can be deprecated upon that work's release.

**Comment by bradzacher** [comment]
It's worth mentioning that rules like `@typescript-eslint/no-unused-vars` rely on typescript-eslint's scope analyser AND the logic in the extension rule. That may be a bit too large a piece of work for this workstream unless you want to look into updating `eslint-scope` as well.

**Comment by nzakas** [comment]
> Switching back from [#19373 (comment)](https://github.com/eslint/eslint/issues/19373#issuecomment-2619979489): I'm not sure we're aligned on what _"TypeScript syntax-aware"_ means here. What are the kinds of changes marked as accepted for these rules?

@JoshuaKGoldberg any changes that can be done by detecting TypeScript-specific AST nodes. This is to differentiate from *type aware*, which would mean needing to understand TypeScript types.

> It's worth mentioning that rules like `@typescript-eslint/no-unused-vars` rely on the logic in the extension rule **_AND_** typescript-eslint's scope analyser.
> Migrating those may be a bit too large a piece of work for this workstream unless you want to look into updating `eslint-scope` as well.

@bradzacher at least for the first go-around, I think we'd like to stick with just changes inside of the rules. I'm not opposed to updating `eslint-scope`, although I'd like a better idea about what type of changes would be required before committing.

**Comment by JoshuaKGoldberg** [comment]
This means we'd have to discuss what the AST shape should be then, right? Whether it's exactly what typescript-eslint currently produces, or some other structure? I think working on any rules is blocked on updating the parser to support those TypeScript AST nodes.

**Comment by nzakas** [comment]
@JoshuaKGoldberg I was thinking we'd just use what typescript-eslint produces. I don't see updating Espree as related to this effort (especially as it would mirror what typescript-eslint produces).

**Comment by nzakas** [comment]
Do we have any volunteers to help get this started?

**Comment by JoshuaKGoldberg** [comment]
#19431 is a first attempt at a strategy.

**Comment by nzakas** [comment]
#19431 made me think: should we add something into rule `meta` when it supports TypeScript syntax?

**Comment by nzakas** [comment]
**TSC Summary:** In this issue, we agreed to start making some core rules TypeScript syntax-aware.

**TSC Question:** Should we add something into the rule `meta` to reflect this? If so, what should it be? And should we take into account that rules may be aware of other languages and/or types as well?

**Comment by sam3k** [comment]
It was agreed on 2025-02-20 TSC meeting to add `meta.language` and `meta.dialects`, starting with rules that are TypeScript-enabled. We can revisit discussions around how to use that data at a later time.

**Comment by snitin315** [comment]
Picking up `init-declarations`

**Comment by SwetaTanwar** [comment]
I am working on [no-unused-expressions](https://github.com/eslint/eslint/pull/19564)

**Comment by davidglezz** [comment]
Hi! I would like to ask for support for this case:

In case of functions with `this` (unboundThis), it should not be autofixed because it generates incorrect code.

https://github.com/typescript-eslint/typescript-eslint/issues/11058

**Comment by Tanujkanti4441** [comment]
> Hi! I would like to ask for support for this case:
> 
> In case of functions with `this` (unboundThis), it should not be autofixed because it generates incorrect code.
> 
> [typescript-eslint/typescript-eslint#11058](https://github.com/typescript-eslint/typescript-eslint/issues/11058)

That sounds reasonable. I'll send a PR for this.

**Comment by nzakas** [comment]
Just a note (@davidglezz and everyone else), this isn't the right place to ask for bug fixes or enhancements. Please open separate issues so we can triage and track them. This issue is just for adding initial TypeScript support to core rules.

**Comment by joealden** [comment]
Could/should this apply to [`no-restricted-globals`](https://eslint.org/docs/latest/rules/no-restricted-globals)? The rule doesn't break when linting TS code from what I can tell, but it doesn't explicitly "support" TS as far as I'm aware from looking at the code + the git blame. Like https://github.com/typescript-eslint/typescript-eslint/issues/4715 mentions, it'd be great if globals that are used in type contexts could be ignored/have different logic applied.

**Comment by tetarchus** [comment]
@JoshuaKGoldberg - Been running through the typescript rules for a config and came across [`no-unused-private-class-members`](https://typescript-eslint.io/rules/no-unused-private-class-members) which seems to fit the bill for this (it isn't marked as needing type information, and extends an existing rule), but doesn't appear in the tracker.

Should this rule be included here?

**Comment by JoshuaKGoldberg** [comment]
Good catch! Yes, I think it should be. Does anybody from the team have any objections?

**Comment by bradzacher** [comment]
I didn't build this into ESLint core cos I thought it might be a bit too "typescript-specific" and there were a number of edge cases with it that couldn't be handled easily as TS `private` members don't have the same syntactic restrictions that `#private` members do.

The implementation could be rewritten into eslint core and pure JS easy enough so long as they're okay taking on the limitations.
